src/Orleans/Utils/Factory.cs

```diff
@@ -0,0 +1,36 @@
+﻿namespace Orleans
+{
+    /// <summary>
+    /// Creates an instance of <typeparamref name="TInstance"/>.
+    /// </summary>
+    /// <typeparam name="TInstance"></typeparam>
+    /// <returns></returns>
+    public delegate TInstance Factory<out TInstance>();
```

Comment:
>"overloads"

Understood, example factory was just intended to communicate the idea, factories with args are possible.

> The functional approach also covers each of the items you listed.

In an absolute sense, yes, but not, imo, in an OO intuitive way.  By passing factories as function, we're exposing this capability in a functional pattern where state is not expected.  That is, it is not expected, or arguably correct, for the factories to have state.  By having the factories exposed clearly as objects themselves, it is understood that they are allowed to be statefull.